File: apps/android/app/src/main/java/dev/tohure/android_rust_test/adapter/ContractMessages.kt
package dev.tohure.android_rust_test.adapter

import dev.tohure.android_rust_test.contract.MessageSource
import uniffi.core_financiero.DomainException

/**
 * El nombre que `contracts/cases.json` le da a este error.
 *
 * `DomainError::contract_name()` es un método de Rust y **no cruza el FFI**: el enum
 * generado trae solo los nombres en inglés. Verificado en la Fase 1: los nueve nombres
 * del contrato aparecen 0 veces en el `.kt` generado.
 *
 * El `when` va **exhaustivo, como expresión y sin rama `else`**. Es deliberado: agregar
 * una décima variante al core tiene que romper la compilación acá — un fallo ruidoso y
 * ubicado— en vez de caer en un `"Desconocido"` que compila, pasa en verde, y se descubre
 * el día de la demo cuando esta app muestra un error que las otras tres no.
 */
fun DomainException.contractName(): String =
    when (this) {
        is DomainException.Length -> "Longitud"
        is DomainException.CheckDigit -> "DigitoControl"
        is DomainException.UnknownBank -> "BancoDesconocido"
        is DomainException.InvalidAmount -> "MontoInvalido"
        is DomainException.AccountNotFound -> "CuentaNoEncontrada"
        is DomainException.SameAccount -> "MismaCuenta"
        is DomainException.InsufficientFunds -> "SaldoInsuficiente"
        is DomainException.Encryption -> "Cifrado"
        is DomainException.OutOfRange -> "FueraDeRango"
    }

/**
 * Traduce un error del core al texto que ve el usuario.
 *
 * Vive en producción, no solo en el test, porque `contracts/messages.es.json` indexa los
 * mensajes por nombre del contrato: la pantalla de error necesita el mapeo igual que el
 * golden. **El golden reusa esta misma función** en vez de escribir la suya — así verifica
 * contra `cases.json` el mapeo que la UI usa de verdad, y no una copia que puede divergir.
 */
class ContractMessages(source: MessageSource) {
    private val messages: Map<String, String> = source.messages()

    fun userMessage(e: DomainException): String {
        val name = e.contractName()
        val template = messages[name]
            ?: error("contracts/messages.es.json no tiene el mensaje de `$name`")
        return interpolate(template, e)
    }

    /**
     * Reemplaza `{code}`, `{id}`, `{available}`, `{required}` y `{field}` por los campos
     * de la variante, **crudos**. Nada de `NumberFormat` acá: los formateadores de moneda
     * de Android, iOS y el navegador no coinciden entre sí, y una diferencia rompe la
     * comparación carácter por carácter. El formateo vive en las pantallas de montos.
     */
    private fun interpolate(template: String, e: DomainException): String =
        when (e) {
            is DomainException.Length -> template
            is DomainException.CheckDigit -> template
            is DomainException.UnknownBank -> template.replace("{code}", e.code)
            is DomainException.InvalidAmount -> template
            is DomainException.AccountNotFound -> template.replace("{id}", e.id)
            is DomainException.SameAccount -> template
            is DomainException.InsufficientFunds ->
                template
                    .replace("{available}", e.available)
                    .replace("{required}", e.required)
            is DomainException.Encryption -> template
            is DomainException.OutOfRange -> template.replace("{field}", e.`field`)
        }
}

File: apps/ios/ios-rust-test/Adapter/ContractMessages.swift
import Foundation
import OSLog

extension DomainError {
    /// El nombre que `contracts/cases.json` le da a este error.
    ///
    /// `DomainError::contract_name()` es un método de Rust y **no cruza el FFI**: el enum
    /// generado trae solo los nombres en inglés.
    ///
    /// El `switch` va **exhaustivo y sin `default`**. Es deliberado: agregar una undécima
    /// variante al core tiene que romper la compilación aquí —un fallo ruidoso y ubicado—
    /// en vez de caer en un `"Desconocido"` que compila, pasa en verde, y se descubre el
    /// día de la demo cuando esta app muestra un error que las otras tres no.
    var contractName: String {
        switch self {
        case .Length: return "Longitud"
        case .CheckDigit: return "DigitoControl"
        case .UnknownBank: return "BancoDesconocido"
        case .InvalidAmount: return "MontoInvalido"
        case .AccountNotFound: return "CuentaNoEncontrada"
        case .SameAccount: return "MismaCuenta"
        case .InsufficientFunds: return "SaldoInsuficiente"
        case .Encryption: return "Cifrado"
        case .Decryption: return "Descifrado"
        case .OutOfRange: return "FueraDeRango"
        }
    }
}

/// Traduce un error del core al texto que ve el usuario.
///
/// Vive en producción y no solo en el test porque `contracts/messages.es.json` indexa los
/// mensajes por nombre del contrato: la pantalla de error necesita el mapeo igual que el
/// test de contrato. **El test de contrato reusa esta misma función** en vez de escribir
/// la suya, así que verifica contra `cases.json` el mapeo que la UI usa de verdad.
struct ContractMessages {
    private let table: [String: String]
    private let logger = Logger(subsystem: "dev.tohure.ios-rust-test", category: "CoreFinanciero")

    init(source: MessageSource) {
        table = source.messages()
    }

    /// Lo que ve el usuario cuando falla algo que **no** es un error de dominio. Normativo en
    /// `docs/ui-spec.md`, igual en las cuatro apps. No dice «vuelve a intentarlo» a propósito:
    /// si el core no responde, reintentar no arregla nada, y prometer una salida que no existe
    /// es peor que no decir nada.
    static let fallback = "No se pudo completar la operación."

    /// El texto para un `Error` cualquiera, que es lo que de verdad llega al `catch` genérico
    /// de los ViewModels.
    ///
    /// **No es defensivo, cubre un caso real.** El `catch` final de cada ViewModel atrapa
    /// cualquier cosa que no sea `DomainError`, y antes guardaba `"\(error)"` — o sea el texto
    /// de diagnóstico, en la cara del usuario. En Android el equivalente mostraba
    /// `java.lang.UnsatisfiedLinkError: dlopen failed: …`, verificado con un test.
    ///
    /// **El diagnóstico no se pierde: se loguea acá.** En Android se loguea en el adapter,
    /// porque allá `runCatching` es el único punto donde se atrapa; acá no hay tal punto —los
    /// métodos del protocolo son `throws` y cada ViewModel tiene su `catch`—, así que el lugar
    /// con un solo dueño es éste.
    func userMessage(_ error: Error) -> String {
        if let domain = error as? DomainError {
            return userMessage(domain)
        }
        logger.error("error no-dominio cruzando el FFI: \(String(describing: error))")
        return Self.fallback
    }

    func userMessage(_ error: DomainError) -> String {
        let name = error.contractName
        guard let template = table[name] else {
            return "Error del núcleo: \(name)"
        }
        return interpolate(template, error)
    }

    /// Reemplaza `{code}`, `{id}`, `{available}`, `{required}` y `{field}` por los campos de
    /// la variante, **crudos**. Nada de `NumberFormatter` aquí: los formateadores de Android,
    /// iOS y el navegador no coinciden entre sí, y una diferencia rompe la comparación
    /// carácter por carácter que es toda la tesis. El formateo vive en las pantallas.
    ///
    /// `Longitud`, `MontoInvalido`, `Cifrado` y `Descifrado` llevan campos
    /// (`field`/`expected`/`received`, `detail`) que `contracts/messages.es.json` no usa en su
    /// plantilla — ver `contracts/README.md` — así que esos cuatro casos devuelven la plantilla
    /// sin tocar, igual que `CheckDigit` y `SameAccount`, que no llevan campos.
    private func interpolate(_ template: String, _ error: DomainError) -> String {
        switch error {
        case .Length, .CheckDigit, .InvalidAmount, .SameAccount, .Encryption, .Decryption:
            return template
        case .UnknownBank(let code):
            return template.replacingOccurrences(of: "{code}", with: code)
        case .AccountNotFound(let id):
            return template.replacingOccurrences(of: "{id}", with: id)
        case .InsufficientFunds(let available, let required):
            return
                template
                .replacingOccurrences(of: "{available}", with: available)
                .replacingOccurrences(of: "{required}", with: required)
        case .OutOfRange(let field):
            return template.replacingOccurrences(of: "{field}", with: field)
        }
    }
}
